### PR TITLE
arch/riscv: Add mtimer driver

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
Could be used with:

1. up_alarm_set_lowerhalf to become a scheduler timer
2. nxsched_oneshot_extclk to become a loading timer

## Impact
New driver

### CODE REVIEW

File: arch/risc-v/src/common/riscv_mtimer.c [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+      priv->mtime     = mtime;
+      priv->mtimecmp  = mtimecmp;
+      priv->freq      = freq;
+
+      riscv_mtimer_set_mtimecmp(priv, UINT64_MAX);
+      irq_attach(irq, riscv_mtimer_interrupt, priv);
+      up_enable_irq(irq);
+    }
+
+  return &priv->lower;
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_mtimer.c:305]: So i case if `kmm_zalloc` we will de-reference NULL pointer?

**Comment by pkarashchenko** [comment]
Are you going to add file to compilation list and enable in one of the configs?

**Comment by xiaoxiang781216** [comment]
> Are you going to add file to compilation list and enable in one of the configs?

@no1wudi could enable it on bl602.

File: arch/risc-v/src/common/riscv_mtimer.c (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_mtimer.c:305]: > So i case if `kmm_zalloc` we will de-reference NULL pointer?

There isn't de-reference(*) operation. But, let me change to the cast which is more clear.

Comment (xiaoxiang781216) [reply arch/risc-v/src/common/riscv_mtimer.c]: Done.

Comment (pkarashchenko) [reply arch/risc-v/src/common/riscv_mtimer.c]: Sorry, I missed that `lower` is the first element in the struct. Indeed compiler will eliminate de-reference operation in such case. But anyway now it is more clear without need to go into details. Thank you

**Review by pkarashchenko — APPROVED** [review]

**Review by pussuw — APPROVED** [review]
LGTM. As a ticker it will have a bit of bias from using mtime (current time) instead of mtimecmp (the last alarm time) as seed for the next alarm. But having a generic alarm timer is more worth it than just having a ticker.

**Pull request merged**